### ISSUE TITLE
GPU memory keeps growing and finally reports OOM error

### ISSUE THREAD

**Issue by SunDoge** [NVIDIA/DALI#328]
I ran the [pytorch exmaple](https://github.com/NVIDIA/DALI/blob/master/docs/examples/pytorch/main.py) with `--batch_size=512 --lr=0.2` on ImageNet. The GPU memory usage became higher after each epoch. I notice this problem when I was using `DALI v0.4.1`. I upgrade to `v0.5.0` today and  the problem still exists.

| Epoch | Memory-Usage      |
|-------|-------------------|
| 1     | 9713MiB/12196MiB  |
| 2     | 10845MiB/12196MiB |
| 3     | forgotten         |
| 4     | 11133MiB/12196MiB |

Using 4 `TITAN X (Pascal)`, Driver Version: 396.54

**Comment by SunDoge** [comment]
I use `nn.DataParallel` so actually every GPU is trained with batch size 128. The main GPU loading data consumes over 100000MiB memory and the rest three GPU use about 4000MiB. I can use smaller batch size but the question is that after every epoch the memory consumption grows about 100MiB, which may be a memory leak.

I don't know how to debug it, could you provide me some advice?

**Comment by JanuszL** [comment]
Hi,
The batch size you provide to DALI is not the total one, but per GPU, so when you have 4 GPU then dataset consumed at each iteration is `batch_size * 4`, so if you want to go with 512 (in total) please use `--batch_size=128`. Also, DALI allocates some memory during the run (to adjust internal buffers to data it loads during the training) and as soon as it hits watermark, memory consumption should stay on the same level.

**Comment by SunDoge** [comment]
Hi. I recently notice that when I place `Pipeline` instance on a single GPU, the memory usage is quite stable. I wonder if it is really a bug. Maybe it is caused by other process on the same GPU.

**Comment by Kh4L** [comment]
Hi @SunDoge 
What do you mean exactly by placing `Pipeline` instance on a single GPU?
Running with a World size of 1, or changing `device_id=args.local_rank` to `device_id=0` in the `Pipeline` constructor?

**Comment by SunDoge** [comment]
@Kh4L  Change `device_id` to an idle GPU with no other process running on it. It has been running for two days and the memory usage is quite constant.

**Comment by JanuszL** [comment]
> @Kh4L Change `device_id` to an idle GPU with no other process running on it. It has been running for two days and the memory usage is quite constant.
@SunDoge - could you provide us more details, how exactly you run your training? What HW configuration you have (GPUs and their IDs in the system)? Because I don't fully understand how you run your training now.

**Comment by SunDoge** [comment]
The `Pipelines`:
[https://github.com/SunDoge/ShuffleNetV2-pytorch/blob/3-7-3/data_dali.py](https://github.com/SunDoge/ShuffleNetV2-pytorch/blob/3-7-3/data_dali.py)

I run it with:
```
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5 python imagenet_dali.py \
--dataroot /home/datasets/imagenet \
-j 4 \
--scaling 0.5 \
-b 1024 \
--clr \
--max-lr 1 \
--gpu 0,1,2,3
```
I set the `device_id` to 4 and 5 because I want to leave enough space for it to leak.

The hardware information is shown below:
```
Mon Dec 10 18:56:34 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 396.54                 Driver Version: 396.54                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  TITAN X (Pascal)    Off  | 00000000:04:00.0 Off |                  N/A |
| 31%   54C    P2    94W / 250W |  10128MiB / 12196MiB |      1%      Default |
+-------------------------------+----------------------+----------------------+
|   1  TITAN X (Pascal)    Off  | 00000000:05:00.0 Off |                  N/A |
| 31%   54C    P2   203W / 250W |   9110MiB / 12196MiB |      1%      Default |
+-------------------------------+----------------------+----------------------+
|   2  TITAN X (Pascal)    Off  | 00000000:08:00.0 Off |                  N/A |
| 25%   45C    P2   218W / 250W |   9110MiB / 12196MiB |      1%      Default |
+-------------------------------+----------------------+----------------------+
|   3  TITAN X (Pascal)    Off  | 00000000:09:00.0 Off |                  N/A |
| 33%   56C    P2   294W / 250W |   9110MiB / 12196MiB |     48%      Default |
+-------------------------------+----------------------+----------------------+
|   4  TITAN X (Pascal)    Off  | 00000000:83:00.0 Off |                  N/A |
| 27%   47C    P2    67W / 250W |   6343MiB / 12196MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   5  TITAN X (Pascal)    Off  | 00000000:84:00.0 Off |                  N/A |
| 23%   32C    P8    17W / 250W |   5555MiB / 12196MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   6  TITAN X (Pascal)    Off  | 00000000:87:00.0 Off |                  N/A |
| 26%   48C    P2   199W / 250W |   7884MiB / 12196MiB |     99%      Default |
+-------------------------------+----------------------+----------------------+
|   7  TITAN X (Pascal)    Off  | 00000000:88:00.0 Off |                  N/A |
| 34%   55C    P2    79W / 250W |   7884MiB / 12196MiB |     88%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     10992      C   python                                      5421MiB |
|    0     11554      C   python                                      4671MiB |
|    1     10992      C   python                                      4861MiB |
|    1     11554      C   python                                      4213MiB |
|    2     10992      C   python                                      4861MiB |
|    2     11554      C   python                                      4213MiB |
|    3     10992      C   python                                      4861MiB |
|    3     11554      C   python                                      4213MiB |
|    4     11554      C   python                                      6331MiB |
|    5     11554      C   python                                      5543MiB |
|    6     10992      C   python                                      4861MiB |
|    6     15652      C   python                                      2987MiB |
|    7     10992      C   python                                      4861MiB |
|    7     17923      C   python                                      2987MiB |
+-----------------------------------------------------------------------------+

```

**Comment by JanuszL** [comment]
@SunDoge  - I see that you transfer data between devices inside your training script. DALI works on GPU 4 and 5 and then you transfer it to the remaining one. How we use DALI you can find in https://github.com/NVIDIA/DALI/blob/master/docs/examples/pytorch/renet50/pytorch-resnet50.rst. Basically DALI is instantiated for each GPU separately and data doesn't move between GPUs. Also batch size is set per GPU, not globally. 
It is only my curiosity - did you check that is the penalty of moving data across the GPU?

**Comment by SunDoge** [comment]
So I should write `distributed` code instead of wrapping model in `nn.DataParallel`, right? I get use to use `DataParallel` in `pytorch` because it's much simpler. I'll try it.

**Comment by JanuszL** [comment]
@SunDoge - I'm not saying this is wrong, but  recommended way to obtain the best performance with our GPU training is to use 'distributed' strategy - please check post from csarofeen https://discuss.pytorch.org/t/what-is-the-difference-between-dataparallel-and-distributeddataparallel/6108.
